### PR TITLE
Update lib to work with prices/django-prices beta 3

### PR DESCRIPTION
**Pull request by rafalp**
This PR is followup for changes that occured in prices 1.0.0b3 and django-prices. In addition this PR adds extra tests for better coverage as well as synces our requirements and tests config with prices and django prices.

**TODO**

* [x] Make tests pass
* [x] Sync requirements with other libs
* [x] Test `discount_amount_in_currency`
* [x] Rename 'value' and 'price' to 'base' whenever value's usage is unknown or too generic
* [x] Remove `from __future__` imports
* [x] Try refactoring `convert_price`
* [x] Add `mypy` type annotations wherever possible
* [x] Remove `prices_multicurrency_i18n` filters
* [x] Remove `discount_amount_in_currency` filter
* [x] Cut down number of times that  `exchange_currency` fetches conversion rate for types other than `Money`

### CODE REVIEW

File: django_prices_openexchangerates/__init__.py [hunk 1]
```diff
@@ -16,7 +16,7 @@
 class CurrencyConversion(object):
```
Comment (patrys) [inline django_prices_openexchangerates/__init__.py]: In Python 3 `(object)` is no more.

File: django_prices_openexchangerates/__init__.py [hunk 2]
```diff
@@ -29,14 +29,14 @@ def __repr__(self):
...
-        if isinstance(price_obj, Price):
-            return Price(Amount(price_obj.net.value * self.rate,
-                                currency=self.to_currency),
-                         Amount(price_obj.gross.value * self.rate,
-                                currency=self.to_currency))
+        if isinstance(price_obj, Money):
+            return Money(
+                price_obj.amount * self.rate, currency=self.to_currency)
+        if isinstance(price_obj, TaxedMoney):
+            return TaxedMoney(Money(price_obj.net.amount * self.rate,
```
Comment (patrys) [inline django_prices_openexchangerates/__init__.py]: Please use normal wrap after the opening parenthesis.

File: django_prices_openexchangerates/templatetags/prices_multicurrency_i18n.py [hunk 3]
```diff
@@ -8,18 +8,18 @@
...
+    return converted_value
 
 
 @register.simple_tag
-def discount_amount_in_currency(discount, amount, currency):
-    amount = exchange_currency(amount, to_currency=currency)
-    discount_amount = exchange_currency(discount.amount, to_currency=currency)
-    discount.amount = discount_amount
-    return (amount | discount) - amount
+def discount_amount_in_currency(discount, price, currency):
```
Comment (patrys) [inline django_prices_openexchangerates/templatetags/prices_multicurrency_i18n.py]: I think `base` is better than `price`. Not sure whether we should keep this order of arguments.

File: django_prices_openexchangerates/__init__.py [hunk 4]
```diff
@@ -53,7 +53,7 @@ def get_conversion_rate(currency):
 
 def convert_price(price, to_currency, get_rate=get_conversion_rate):
```
Comment (patrys) [inline django_prices_openexchangerates/__init__.py]: What is a `price` here (both in function name and in signature)?

File: django_prices_openexchangerates/templatetags/prices_multicurrency_i18n.py (hunk 3 above)
Comment (rafalp) [reply django_prices_openexchangerates/templatetags/prices_multicurrency_i18n.py]: The arguments here are mostly like that for consistency with `django-prices`'s `discount_amount_for(discount, price)` template tag.

**Comment by rafalp** [comment]
@patrys I've tried moving `convert_price`'s logic  around. I'm still unsure if we want to keep the `CurrencyConversion` around, or perhaps extract its `apply` logic to function/move it to `exchange_currency`.

**Comment by patrys** [comment]
I would drop it for the same reason we've dropped discount and tax classes. It's just as easy to use `functools.partial` to achieve the same effect.

**Comment by rafalp** [comment]
Removed the `CurrencyConversion` and moved it's apply logic to `exchange_currency`. Seeing how we are py3k only now, I've also removed `__future__` form codebase. If we are on the mark here, I'll move on to adding type annotations for `mypy`.

File: django_prices_openexchangerates/__init__.py [hunk 5]
```diff
@@ -1,45 +1,19 @@
...
 import operator
 from decimal import Decimal
+from typing import Callable, TypeVar, Union
 
 from django.conf import settings
-from prices import Amount, Price, PriceRange
+from prices import Money, MoneyRange, TaxedMoney, TaxedMoneyRange
 
-BASE_CURRENCY = getattr(settings, 'OPENEXCHANGERATES_BASE_CURRENCY', 'USD')
+Conversion = Union[float, int, Decimal]
```
Comment (patrys) [inline django_prices_openexchangerates/__init__.py]: Are you sure we want to support `float`?

File: django_prices_openexchangerates/__init__.py [hunk 6]
```diff
@@ -72,22 +45,38 @@ def convert_price(price, to_currency, get_rate=get_conversion_rate):
...
+        # Exchange to base currency first
+        base = exchange_currency(base, BASE_CURRENCY, get_rate=get_rate)
+
+    conversion_rate = get_conversion_rate(base.currency, to_currency, get_rate)
+
+    if isinstance(base, Money):
+        return Money(base.amount * conversion_rate, currency=to_currency)
+    if isinstance(base, MoneyRange):
+        return MoneyRange(exchange_currency(base.start, to_currency,
+                                            get_rate=get_rate),
```
Comment (patrys) [inline django_prices_openexchangerates/__init__.py]: Please don't do line continuation like that. Fold after each opening parenthesis.

File: django_prices_openexchangerates/__init__.py [hunk 7]
```diff
@@ -72,22 +45,38 @@ def convert_price(price, to_currency, get_rate=get_conversion_rate):
...
+    if isinstance(base, TaxedMoney):
+        return TaxedMoney(Money(base.net.amount * conversion_rate,
+                                currency=to_currency),
+                          Money(base.gross.amount * conversion_rate,
+                                currency=to_currency))
+    if isinstance(base, TaxedMoneyRange):
+        return TaxedMoneyRange(exchange_currency(base.start, to_currency,
+                                                 get_rate=get_rate),
+                               exchange_currency(base.stop, to_currency,
+                                                 get_rate=get_rate))
```
Comment (patrys) [inline django_prices_openexchangerates/__init__.py]: What if none of the conditionals match?

File: django_prices_openexchangerates/migrations/0001_initial.py [hunk 8]
```diff
@@ -1,17 +1,12 @@
 # -*- coding: utf-8 -*-
 # Generated by Django 1.9.2 on 2016-03-17 09:36
-from __future__ import unicode_literals
```
Comment (patrys) [inline django_prices_openexchangerates/migrations/0001_initial.py]: Are you sure that editing migrations is a good idea?

File: django_prices_openexchangerates/templatetags/prices_multicurrency_i18n.py [hunk 9]
```diff
@@ -1,27 +1,32 @@
...
 
 
 @register.simple_tag
-def discount_amount_in_currency(discount, amount, currency):
-    amount = exchange_currency(amount, to_currency=currency)
-    discount_amount = exchange_currency(discount.amount, to_currency=currency)
-    discount.amount = discount_amount
-    return (amount | discount) - amount
+def discount_amount_in_currency(
+    base: T, discount: Callable[[T], T], currency: str) -> T:
```
Comment (patrys) [inline django_prices_openexchangerates/templatetags/prices_multicurrency_i18n.py]: Bad indentation.

File: setup.py [hunk 10]
```diff
@@ -33,7 +32,8 @@
         'django_prices_openexchangerates.templatetags'],
     include_package_data=True,
     classifiers=CLASSIFIERS,
-    install_requires=['Django>=1.4', 'django-prices>=1.0.0-beta2', 'prices>=1.0.0-beta2'],
+    install_requires=[
+        'Django<2.1.0', 'django-prices>=1.0.0-beta3', 'prices>=1.0.0-beta3'],
```
Comment (patrys) [inline setup.py]: Why `Django<2.1.0` instead of `Django>=1.11`?

File: django_prices_openexchangerates/migrations/0001_initial.py (hunk 8 above)
Comment (rafalp) [reply django_prices_openexchangerates/migrations/0001_initial.py]: I've tested this change locally with Django 2.0.1 on py35k and py36k and can confirm that `makemigrations` reports `No changes detected` after the removal of `__future__` import, something I'm consider farm more acceptable than editing `operations` content.

File: django_prices_openexchangerates/__init__.py [hunk 11]
```diff
@@ -1,45 +1,19 @@
...
-                price_obj.value * self.rate, currency=self.to_currency)
-        if isinstance(price_obj, Price):
-            return Price(Amount(price_obj.net.value * self.rate,
-                                currency=self.to_currency),
-                         Amount(price_obj.gross.value * self.rate,
-                                currency=self.to_currency))
-
-
-def get_conversion_rate(currency):
+def get_rate_from_db(currency: str) -> Conversion:
```
Comment (patrys) [inline django_prices_openexchangerates/__init__.py]: Why does it return a union here? Does the returned type differ between calls?

File: django_prices_openexchangerates/__init__.py [hunk 12]
```diff
@@ -51,43 +25,57 @@ def get_conversion_rate(currency):
     return rate.rate
 
 
-def convert_price(price, to_currency, get_rate=get_conversion_rate):
+def get_conversion_rate(
+        from_currency: str, to_currency: str,
+        get_rate: Callable[[str], Conversion]) -> Conversion:
```
Comment (patrys) [inline django_prices_openexchangerates/__init__.py]: Same, why return a union type?

File: django_prices_openexchangerates/__init__.py [hunk 13]
```diff
@@ -51,43 +25,57 @@ def get_conversion_rate(currency):
...
+
+    if isinstance(base, Money):
+        return Money(base.amount * conversion_rate, currency=to_currency)
+    if isinstance(base, MoneyRange):
+        return MoneyRange(
+            exchange_currency(base.start, to_currency, get_rate=get_rate),
+            exchange_currency(base.stop, to_currency, get_rate=get_rate))
+    if isinstance(base, TaxedMoney):
+        return TaxedMoney(
+            Money(base.net.amount * conversion_rate, currency=to_currency),
```
Comment (patrys) [inline django_prices_openexchangerates/__init__.py]: A recursive call would make it more readable.

File: django_prices_openexchangerates/templatetags/prices_multicurrency.py [hunk 14]
```diff
@@ -1,19 +1,23 @@
...
 
 
 @register.filter
-def in_currency(amount, currency, **kwargs):
-    converted_amount = exchange_currency(amount, currency)
-    converted_amount = converted_amount.quantize('.01')
-    return converted_amount
+def in_currency(base: T, currency: str) -> T:
+    converted_base = exchange_currency(base, currency)
+    converted_base = converted_base.quantize('.01')
```
Comment (patrys) [inline django_prices_openexchangerates/templatetags/prices_multicurrency.py]: This quantize call is wrong. We can't assume two decimal places.

File: django_prices_openexchangerates/templatetags/prices_multicurrency_i18n.py [hunk 15]
```diff
@@ -1,27 +1,32 @@
...
 
 
 @register.filter
-def in_currency(amount, currency, **kwargs):
-    converted_amount = exchange_currency(amount, currency)
-    converted_amount = converted_amount.quantize('.01')
-    return converted_amount
+def in_currency(base: T, currency: str) -> T:
+    converted_base = exchange_currency(base, currency)
+    converted_base = converted_base.quantize('.01')
```
Comment (patrys) [inline django_prices_openexchangerates/templatetags/prices_multicurrency_i18n.py]: Also wrong.

File: django_prices_openexchangerates/templatetags/prices_multicurrency_i18n.py [hunk 16]
```diff
@@ -1,27 +1,32 @@
...
+    converted_base = exchange_currency(base, to_currency=currency)
+    discount_amount = exchange_currency(discount_amount, to_currency=currency)
+    return discount_amount - converted_base
 
 
 @register.filter
-def amount(obj, format='text'):
-    return prices_i18n.amount(obj, format=format)
+def amount(base: Money, format='text') -> str:
+    return prices_i18n.amount(base, format=format)
```
Comment (patrys) [inline django_prices_openexchangerates/templatetags/prices_multicurrency_i18n.py]: Why is this templatetag useful?

**Comment by rafalp** [comment]
Another pass on our codebase:

- Removed `discount_amount_in_currency` because it didn't fit into mental image of what lib should do (currency conversion, not math)
- Removed `i18n` version of templatetags because its codebase was identical to non-i18n module
- Recurrent calls inside `exchange_currency` reuse `conversion_rate`, greatly shaving number of hits to outside resources for complex money types like `TaxedMoneyRange`.
- Cleanup the tests suite from tests that claimed to test different things but in fact tested same inputs and results.
- Address some style issues raised by `pycodestyle`, such as double blank lines in otherwhise empty files

**Pull request merged**